### PR TITLE
Dev/reg abstraction

### PR DESCRIPTION
**Pull request by jelavice**
This is a first step towards abstracting away the scan matching. I will sort out the rest of the parameters and the documentation later (too keep the PR not too big).

### CODE REVIEW

File: open3d_slam/include/open3d_slam/Parameters.hpp [hunk 1]
```diff
@@ -26,22 +26,19 @@ enum class IcpObjective : int {
 	PointToPlane
 };
 
-enum class MesherStrategy : int {
-	AlphaShape,
-	BallPivot,
-	Poisson
+static const std::map<std::string, IcpObjective> IcpObjectiveNames {
```
Comment (nubertj) [inline open3d_slam/include/open3d_slam/Parameters.hpp:29]: maybe this variable can have a better name. From the given one one could think it is a vector of strings or so. I often refer to these maps as ..To...

E.g. here it could be objectiveNameToMethodMap

File: open3d_slam/include/open3d_slam/ScanToMapRegistration.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+public:
+	ScanToMapRegistration() = default;
+	virtual ~ScanToMapRegistration() = default;
+	virtual ProcessedScans processForScanMatchingAndMerging(const PointCloud &in,
+			const Transform &mapToRangeSensor) const =0;
+	virtual RegistrationResult scanToMapRegistration(const PointCloud &scan, const Submap &activeSubmap,
+			const Transform &mapToRangeSensor, const Transform &initialGuess) const = 0;
+};
+
+class ScanToMapIcpOpen3D : public ScanToMapRegistration {
```
Comment (nubertj) [inline open3d_slam/include/open3d_slam/ScanToMapRegistration.hpp:40]: ICP seems to be still fairly general for implementing the scanToMapRegistration already. Don't you wanna split it up in the different ICP types at this level?

File: open3d_slam/include/open3d_slam/cloud_processing.hpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/*
```
Comment (nubertj) [inline open3d_slam/include/open3d_slam/cloud_processing.hpp:1]: is this file needed? can it not be made a static method of the Registration base class or so?

**Comment by nubertj** [comment]
In general it works fine for me and the changes make totally sense.

I guess in general some code formatting and slightly more informative variableNames might be helpful, but maybe we can tackle such a code cleaning in the future.

**Review by nubertj — APPROVED** [review]

File: open3d_slam/src/ScanToMapRegistration.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	update(params_);
+}
+void ScanToMapIcpOpen3D::update(const MapperParameters &p) {
+	icpCriteria.max_iteration_ = p.scanMatcher_.maxNumIter_;
+	icpObjective = icpObjectiveFactory(p.scanMatcher_.icpObjective_);
+	mapBuilderCropper_ = croppingVolumeFactory(params_.mapBuilder_.cropper_);
+	scanMatcherCropper_ = croppingVolumeFactory(params_.scanProcessing_.cropper_);
+}
+
+void ScanToMapIcpOpen3D::estimateNormalsIfNeeded(PointCloud *pcl) const {
```
Comment (tutunarsl) [inline open3d_slam/src/ScanToMapRegistration.cpp:37]: Maybe this function can be moved to cloud_processing or with a possible rename: `cloudProcessor` 

Motivation: In the near future we might estimate more than just the normals i.e. curvature, lines, local variance
that way we can just keep adding to this processor class

File: open3d_slam/src/ScanToMapRegistration.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	if (!pcl->HasNormals() && params_.scanMatcher_.icpObjective_ == o3d_slam::IcpObjective::PointToPlane) {
+		estimateNormals(params_.scanMatcher_.kNNnormalEstimation_, pcl);
```
Comment (tutunarsl) [inline open3d_slam/src/ScanToMapRegistration.cpp:39]: Side question: Do we care about surface normal orientation? That consistency might help in other applications such as mesh generation or gradient detection

File: open3d_slam/include/open3d_slam/cloud_processing.hpp (hunk 3 above)
Comment (tutunarsl) [reply open3d_slam/include/open3d_slam/cloud_processing.hpp:1]: The registration base class could be great but I also see value on keeping a `cloud processor` file which might be extended with box filters, retrievers, feature extractors

File: open3d_slam/include/open3d_slam/cloud_processing.hpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ */
+
+#pragma once
+#include <open3d/geometry/PointCloud.h>
+#include "open3d_slam/typedefs.hpp"
+#include <memory>
+
+namespace o3d_slam {
+
+class CroppingVolume;
```
Comment (tutunarsl) [inline open3d_slam/include/open3d_slam/cloud_processing.hpp:15]: Not to be addressed in this PR but I would support for a name like VolumeCropper, but also doesn't affect much

**Review by tutunarsl — COMMENTED** [review]
I think a great start, all the changes makes sense and will help a lot in the future

A question: we currently target the class for ScanToMapRegistration, but could it be a registration base class that ScanToScan also be possible? -> We could set the `Map` as the reference cloud regardless of its size and for ScanToScan we could just set another scan as the reference

Just as a general idea, but quite possibly you had a valid reason to design it this way

**Comment by tutunarsl** [comment]
I agree with @nubertj, I think as it is it can be merged and suggestions can be followed up as separate PRs

**Review by tutunarsl — APPROVED** [review]

File: open3d_slam/include/open3d_slam/cloud_processing.hpp (hunk 3 above)
Comment (jelavice) [reply open3d_slam/include/open3d_slam/cloud_processing.hpp:1]: I can probably remove the separate file. For now we don't need it.

File: open3d_slam/include/open3d_slam/ScanToMapRegistration.hpp (hunk 2 above)
Comment (jelavice) [reply open3d_slam/include/open3d_slam/ScanToMapRegistration.hpp:40]: I did it, but it hasn't been pushed yet. WAnted to keep this PR reasonably sized.

File: open3d_slam/src/ScanToMapRegistration.cpp (hunk 5 above)
Comment (jelavice) [reply open3d_slam/src/ScanToMapRegistration.cpp:39]: Yes, but this can also be reoriented right before you use it. It's just one function call.

File: open3d_slam/include/open3d_slam/cloud_processing.hpp (hunk 6 above)
Comment (jelavice) [reply open3d_slam/include/open3d_slam/cloud_processing.hpp:15]: Yes, that is actually a better name.

File: open3d_slam/src/ScanToMapRegistration.cpp (hunk 4 above)
Comment (jelavice) [reply open3d_slam/src/ScanToMapRegistration.cpp:37]: I added another abstraction for scan-to-scan matching. The scan2map refinement can be tricky if you change the underlying map representation (e.g. you move towards something with voxelized representation and/or voxel hasning, etc.). This is why I kept this the interface where you pass in the Submap class.

Not sure that is the best solution, but it's an improvement compared to what we had earlier where everything was depending on the particular ICP implementation.

Comment (tutunarsl) [reply open3d_slam/src/ScanToMapRegistration.cpp:37]: Thanks for the explanation! This separation is equally beneficial is you already have scan2scan abstraction then forget my previous suggestion :+1:
